### ISSUE TITLE
Some parts throw exception, in debug mode parts cannot be compiled

### ISSUE THREAD

**Issue by Gordon-Dry** [KSPSnark/IndicatorLights#5]
Using a dev build.

The stock battery Squad/Parts/Electrical/z-4kBattery/z-4kBattery/batteryBankLarge cannot be compiled.

In normal game it "only" got no lights.

```
UnityException: get_dataPath is not allowed to be called from a MonoBehaviour constructor (or instance field initializer), call it in Awake or Start instead. Called from MonoBehaviour 'ModuleDataTransmitterIndicator' on game object 'ju1MDA'.
See "Script Serialization" page in the Unity Manual for further details.
  at (wrapper managed-to-native) UnityEngine.Application:get_dataPath ()
  at KSP.IO.IOUtils.get_PluginRootPath () [0x00000] in <filename unknown>:0 
  at KSP.IO.IOUtils.GetFilePathFor (System.Type T, System.String file, .Vessel flight) [0x00000] in <filename unknown>:0 
  at KSP.IO.PluginConfiguration.CreateForType[ModuleControllableEmissive] (.Vessel flight) [0x00000] in <filename unknown>:0 
  at IndicatorLights.Configuration..cctor () [0x00000] in <filename unknown>:0 
Rethrow as TypeInitializationException: An exception was thrown by the type initializer for IndicatorLights.Configuration
  at IndicatorLights.DefaultColorExtensions.Value (DefaultColor defaultColor) [0x00000] in <filename unknown>:0 
  at IndicatorLights.ColorSources+ConstantColorSource..ctor (DefaultColor color) [0x00000] in <filename unknown>:0 
  at IndicatorLights.ColorSources.Constant (DefaultColor color) [0x00000] in <filename unknown>:0 
  at IndicatorLights.ModuleDataTransmitterIndicator..ctor () [0x00000] in <filename unknown>:0 
UnityEngine.GameObject:Internal_AddComponentWithType(Type)
UnityEngine.GameObject:AddComponent(Type) (at C:\buildslave\unity\build\artifacts\generated\common\runtime\GameObjectBindings.gen.cs:382)
Part:AddModule(String, Boolean)
Part:AddModule(ConfigNode, Boolean)
PartLoader:ParsePart(UrlConfig, ConfigNode)
<CompileParts>c__Iterator1:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr) (at C:\buildslave\unity\build\Runtime\Export\Coroutines.cs:17)
 
(Filename:  Line: -1)

```

**Comment by Gordon-Dry** [comment]
PiezPiedPy - today at 03:03
Tell him he needs to do any variable initialization in Awake and not in the constructors

having to use Awake was due to changes when 1.3.1 was released

**Comment by KSPSnark** [comment]
Thanks for the suggestion, but I'm afraid I need more information.

First, to even consider looking at this, I'd need exact repro steps.  Can you clarify what you mean by "debug mode"?  Or by "normal game"?  I have no idea what you mean by either of those things.  When I start up KSP, and build a ship that has a Z4K battery on it, the lights work fine and exactly as I expect, and no exceptions are thrown.

Second, it would also help me to have an idea of impact, e.g. who does this affect?  Certainly it doesn't affect me when I play KSP, and I expect it doesn't affect the vast majority of players, since 1.3.1 has been out for a very long time and nobody's brought it up before now, so it would be helpful to know.  Clarifying "who's affected and how" would be useful to me, since I need to know the impact in order to assess whether and when to address the matter.

Third, I also have to take into account the amount of my time it would take to investigate and fix.  If the time is large relative to the amount of benefit a fix would provide, I probably won't have time to address.  On the other hand, if someone *else* (e.g. someone who cares a lot because they're personally affected) can do the detective work and hand me a specific solution-- i.e. "here's a suggested fix, move this code from this spot to that spot", that sort of thing-- then that greatly reduces the amount of my time required and increases the likelihood of it being practical for me to address.

So, to summarize, 1. please give repro steps, 2. please provide an impact assessment, and 3. if possible, please provide specific advice about a suggested fix ("variable initialization" is pretty vague).  Thanks.

**Comment by Gordon-Dry** [comment]
I installed the development version of KSP to test a new Kerbalism build while I was in Discord.

> How to start KSP in Development (Debug) mode.
> 
> 1. Copy the `KSP_x64_Dbg.exe` file in this zip to your KSP install folder
> 2. Create a junction in your KSP install folder named `KSP_x64_Dbg_Data` linking to your KSP `KSP_x64_Data` folder.
> 
> This is done by opening a command prompt in your KSP install folder and running the following command:
> 
> mklink /J KSP_x64_Dbg_Data KSP_x64_Data
> 
> Now you can choose between the Development version (launch KSP_x64_Dbg.exe)
>  or the released non-Development normal version (run KSP_x64.exe)

I forward your question to PiezPiedPy ...

**Comment by KSPSnark** [comment]
I'm sorry, what exactly is KSP_x64_Dbg.exe, where did you get it from, and why would one use it?

Is this some official thing that Squad has released?  If so, can you provide a link to where one downloads it from?

**Comment by Gordon-Dry** [comment]
When a mod dev also compiles a dll plus debug stuff and debug symbols, it can be used in the debug version of KSP to get a whole lot of more information written to the log.

This is an interesting experience, because mods that barely run with some NRE's but still surviving in normal KSP just commit suicide and take other mods with them in the  debug version.

Debug version is no way mercyful. Bugs just are bugs.

A special debug build of a mod will survive but write a lot to the log.

And also one thing is clearly to see: the core game itself is a hell of a bugfest from the beginning.

The benefit is that in the whole mass of more logging there are informations WHICH mod did WHAT.

In normal KSP the log only says "x happened. Point."

**Comment by Gordon-Dry** [comment]
https://forum.kerbalspaceprogram.com/index.php?/topic/102909-ksp-plugin-debugging-and-profiling-for-visual-studio-and-monodevelop-on-all-os/

PiezPiedPy just gave me a dropbox link where the exe already was in (plus Kerbalism debug build).

**Comment by KSPSnark** [comment]
Thank you, that link gives the needed information.

So, if I'm understanding correctly:  this is an issue that affects people who go through all the steps listed [here](https://forum.kerbalspaceprogram.com/index.php?/topic/102909-ksp-plugin-debugging-and-profiling-for-visual-studio-and-monodevelop-on-all-os/) in order to run KSP in debug mode, but doesn't affect anyone in a regular game (i.e. normal players), correct?

In that case (from my perspective, looking at the cost/benefit analysis of what it would take for me to address this problem), it means that the "impact" is very small-- i.e. take the tiny percentage of people who run KSP in debug mode (tiny enough that it doesn't include me, I've never done this), and then take the small fraction of *those* who also want to run IndicatorLights, and that boils down to very few people indeed.  Then compare that with the many hours of my time it would take to address this, which I don't have available these days.  Add the risk (i.e. that I might break something for the 99.99% of users who aren't running debug mode and are using the current version of IndicatorLights just fine).

So thank you for the suggestion, but unfortunately the balance just doesn't work-- I don't have the bandwidth to try to address something like this.  Small impact, large cost, nonzero risk.

I'll leave the issue open, just in case I ever get into debug-mode KSP programming in the future, but I'm not anticipating doing that any time soon.  I'm sorry it's not practical for me to address this at this time.

**Comment by Gordon-Dry** [comment]
I guess this connects to the issue some ppl mentioned in the forum that on some parts the lights are not working, on other parts they do.

**Comment by PiezPiedPy** [comment]
@Gordon-Dry Exactly correct.

@KSPSnark Your mod should only be built as a release after thoroughly debug testing as is all other software. I recommend you test your mod in Development debug mode to not do so is rather silly.
 It takes a matter of minuets not hours to convert KSP into a development environment. Just use my dropbox link but leave out the Kerbalism stuff.
https://www.dropbox.com/s/8djpj3cs7u15hia/KSP_Devmode.7z?dl=0

**Comment by KSPSnark** [comment]
No plans to fix at this time.
